### PR TITLE
Copy when ast.Modify() so expanding macros expand each macro use as the right one

### LINKED ISSUE #20: ExpandMacros mutates the original program ~~(maybe it's fine)~~ it's not fine :)
Modify should probably clone or have the option to do so

tbd - then maybe just not return a result (outside of errors I guess)

### PR DESCRIPTION
**Pull request by ldemailly**
Fixes:
- #20 
- #223 
- #68

Was a pretty bad bug and #20 was a smoking gun, so where the comments I then wrote in the code
(basically that macros would only work properly "once")

### CODE REVIEW

File: ast/modify_test.go [hunk 1]
```diff
@@ -119,3 +113,29 @@ func TestModify(t *testing.T) {
 		}
 	}
 }
+
+func TestModifyMap(t *testing.T) {
+	// need to test map separately because deepequal can't compare a map
+	// with keys of the same underlying value (2:2, 2:2)
+	modified := Modify(
+		&MapLiteral{
```
Comment (ccoVeille) [inline ast/modify_test.go:121]: DeepEquals cannot, but [maps.Equal](https://pkg.go.dev/maps#Equal) can, no?

File: ast/modify.go [hunk 2]
```diff
@@ -1,55 +1,121 @@
 package ast
 
-import "fortio.org/log"
+import "fmt"
 
 // Note, this is somewhat similar to eval.go's eval... both are "apply"ing.
-func Modify(node Node, f func(Node) Node) Node {
+func Modify(node Node, f func(Node) Node) Node { //nolint:funlen,gocyclo // yeah lots of types.
 	// TODO: add err checks for _s.
```
Comment (ccoVeille) [inline ast/modify.go:7]: What about a map[string]func(node)

Where you use fmt.Sprintf("%T", node) as a key and ask if the node type is in the map, if not, panic, if present call the func?

This way you will have one method per type.

Maybe it's naive, and may need a type assertion in each method, but you could split this gigantic piece of code

File: eval/eval_api.go [hunk 3]
```diff
@@ -170,7 +171,8 @@ func AddEvalResult(name, code string) error {
 func EvalString(this any, code string, emptyEnv bool) (object.Object, error) {
 	l := lexer.New(code)
 	p := parser.New(l)
-	program := p.ParseProgram()
+	var program ast.Node
+	program = p.ParseProgram()
```
Comment (ccoVeille) [inline eval/eval_api.go:175]: I don't get this change

File: ast/modify_test.go (hunk 1 above)
Comment (ldemailly) [reply ast/modify_test.go:121]: == wouldn't be equal as the values (and keys) are pointers to integer

File: eval/eval_api.go (hunk 3 above)
Comment (ldemailly) [reply eval/eval_api.go:175]: ParseProgram returns *ast.Statements which happens to be a node (one of the subtype) but when we call Modify() later it returns a Node and we re-assign it (modified copy) into same variable so it has to be of the right type

File: ast/modify.go (hunk 2 above)
Comment (ldemailly) [reply ast/modify.go:7]: I wouldn't use a string version and keep attaching strings no - now maybe a map can have a type as key, not sure

the code is very mechanical (copy for each type and recurse on each components), a few lines in each case so I don't think the linters alone should be a reason to break it out

File: ast/modify_test.go (hunk 1 above)
Comment (ccoVeille) [reply ast/modify_test.go:121]: Oh indeed.

**Pull request merged**